### ISSUE TITLE
[node-core-library] PackageJsonLookup does not parse versionless package.json files

### ISSUE THREAD

**Issue by shlomiassaf** [Microsoft/web-build-tools#1044]
The method [PackageMetadataManager.isAedocSupportedFor](https://github.com/Microsoft/web-build-tools/blob/master/apps/api-extractor/src/analyzer/PackageMetadataManager.ts#L138) is using `PackageJsonLookup` from `@microsoft/node-core-library` to lookup and load the `package.json` file. (through `PackageMetadataManager.tryFetchPackageMetadata`)

This seems to be safe, as long as the file does not exist.

If the file exists but the [`name` or `version` fields are missing in `package.json` the process will throw](https://github.com/Microsoft/web-build-tools/blob/master/libraries/node-core-library/src/PackageJsonLookup.ts#L200-L207)

A try catch should be in-place, assuming it should be a safe pass.

This is my current workaround, not sure it cover all scenarios:

```ts
import { PackageMetadataManager } from '@microsoft/api-extractor/lib/analyzer/PackageMetadataManager';

const { isAedocSupportedFor } = PackageMetadataManager.prototype;

PackageMetadataManager.prototype.isAedocSupportedFor = function (sourceFilePath: string) {
  try {
    return isAedocSupportedFor.call(this, sourceFilePath);
  } catch (err) {
    return false;
  }
};
```

**Comment by pgonzal** [comment]
Both the the [CommonJS Packages specification](http://wiki.commonjs.org/wiki/Packages/1.0) and [NPM manual](https://docs.npmjs.com/files/package.json) specify that `name` and `version` are *required* fields for the **package.json** file format.

Why are they missing from your **package.json** file?

**Comment by shlomiassaf** [comment]
@octogonz They are not missing from "my" files but from others (quite popular) packages.

You are right, the spec does require `name` and `version` to exists, and they do in all **primary** npm packages...

The problem is with secondary packages, npm packages that come nested inside primary npm pacakges.

For example, with `npm install @angular/material` I get the entire material design component library for angular.

The material package is nothing but a container for a lot of UI components, each bundled as a nested package.

Here is one for example[https://unpkg.com/@angular/material@7.3.0/datepicker/package.json](https://unpkg.com/@angular/material@7.3.0/datepicker/package.json) 

As you can see, the package json is invalid (spec wise) but since it's not the root package it does not have to follow the spec.

Now, we can try and educate the entire world, which will not help us... breaking the process just because of this is to strict...

**Comment by octogonz** [comment]
@shlomiassaf does **@angular/material** have a good reason to publish **package.json** files that do not conform to the spec?  If not, perhaps we should simply open an issue/PR to get Angular fixed.  It's probably a small change.  I don't see why they would object to such a proposal.

My concern is that if our library accepts invalid inputs, every consumer of the library will now need to be updated to check for that edge case and handle it somehow.  This increases the complexity and test surface for our open source projects.  If there is a good reason for omitting the `version` field, then I'm fine with taking on that additional cost. But the reason is merely to accommodation one particular library that doesn't follow the rules, then it seems like we're fixing the wrong end of the problem.

What do you think?

**Comment by shlomiassaf** [comment]
@octogonz It's not just angular/material its also every library inside `@angular` that is **secondary**

Asking them to change **might** help in this case but there will be others...

Let's tackle it from a different angle... The reason there is no **version** is because the version does not exists, not for the secondary library anyway.

You can only download the primary so this is why the primary has a version and the secondary does not.

You might add a check to see the scope of the library (secondary/primary)...

**Comment by octogonz** [comment]
> You can only download the primary so this is why the primary has a version and the secondary does not. > The actual version for the secondary will always be the version of the primary

Right. My point was that Angular's build script should simply copy the primary version into the secondary package.  This should be a small change.  The resulting version is semantically meaningful.

**Comment by shlomiassaf** [comment]
@octogonz It will not work, chasing every other package author to fix stuff will just not work, we need to enable not block.

The fix should be here, enforcing package validity only on primary packages (which will be valid anyway...)

**Comment by octogonz** [comment]
Okay, but there really isn't any concept of a "primary" vs "secondary" package.  NPM allows **require()** to import from any discoverable folder containing a **package.json** file.  So if we wanted to treat "secondary" packages specially, I'm not sure how we would recognize them.

> It will not work, chasing every other package author to fix stuff will just not work, we need to enable not block

Hmm... Could you give examples of other mainstream projects besides Angular that are publishing **package.json** files without a `"version"` field?  If this is indeed a widespread practice, that would be more persuasive.  My suspicion was that this is just Angular-being-Angular(tm). (Google uses an internal tooling ecosystem that is fairly different from the rest of the world, so they sometimes have friction when trying to interact with OSS community projects. For example I heard somewhere that Angular's code is officially tracked in non-Git version control system, and the GitHub repo is a carefully synchronized mirror of it.)  But it's just speculation, maybe I'm wrong.  :-D

**Comment by shlomiassaf** [comment]
> NPM allows **require()** to import from any discoverable folder containing a **package.json** file

NPM does now allow that, this is actually a **node** implementation of modules. NPM will only validate the primary **package.json** and fail on publish if version is missing.

In the [node documentation](https://nodejs.org/docs/latest-v9.x/api/modules.html#modules_folders_as_modules) for modules it is clearly specified the the only requirement is the 'main' property.

![image](https://user-images.githubusercontent.com/5377501/52345054-fc348380-2a24-11e9-9088-bed2287287e5.png)

Note the text: **This is the extent of Node.js's awareness of package.json files.**

**Comment by octogonz** [comment]
That's persuasive. Let's fix it.

**Comment by octogonz** [comment]
Thanks @alan-agius4  for making a PR.

@shlomiassaf  When I tested this, it actually breaks a lot of stuff.  I'm pushing an update that reintroduces an interface `IPackageJsonWithVersion` to avoid having to constantly validated the `version` field in places where it's guaranteed to exist.  It's technically a breaking major API change for **node-core-library**, although we can probably get away with releasing it as SemVer minor.

**Comment by octogonz** [comment]
I have updated the PR so that `IPackageJson` is reverted to its original meaning, and a new `INodePackageJson` is introduced for the case where `version` may be omitted.  This avoids breaking existing code, and I believe it better describes the real world usage of these files. (In particular, I still feel that Angular's being all mavericky, even if what they're doing is technically permitted.)

So now the important thing is to validate that with this change, @shlomiassaf 's original scenario is now accepted by API Extractor.

**Comment by iclanton** [comment]
This has been released with api-extractor beta version [7.0.29](https://www.npmjs.com/package/@microsoft/api-extractor/v/7.0.29)
